lib/ansible/galaxy/token.py

```diff
@@ -35,6 +48,93 @@
 
 display = Display()
 
+AH_TOKEN_EXPIRES_N_MIN_EARLY = 1
+AH_AUTH_URL = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
+AH_JWKS_URI = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/certs"
+
+# https://www.rfc-editor.org/rfc/rfc7518#section-3.1
+REQUIRED_JWT_HEADER_KEYS = set(['alg', 'typ', 'kid'])
+REQUIRED_JWT_HEADER_VALUES = {
+    'alg': ['RS256'],
+    'typ': ['JWT'],
+}
+
+PublicNumbers = namedtuple("PublicNumbers", ["e", "n"])
+
+
+def validate_jwt_format(data):
+    if len(data.split('.')) != 3:
+        raise InvalidSignature(f"unsupported format: {data}")
+
+    if any(whitespace in data.rsplit('.')[0] for whitespace in string.whitespace):
+        raise InvalidSignature(f"JWS header and payload should not contain line breaks or whitespace: {data.split('.')[0]}")
+
+
+def validate_jwt_header(header):
+    if (
+        set(header) != REQUIRED_JWT_HEADER_KEYS
+        or any(header[key] not in value for key, value in REQUIRED_JWT_HEADER_VALUES.items())
+    ):
+        raise InvalidSignature(f"unsupported header: {header}")
+
+
+def b64url_to_long(data):
+    decoded_bytes = base64.urlsafe_b64decode(
+        to_bytes(data, errors="surrogate_or_strict") + b"=="
+    )
+    return int(decoded_bytes.hex(), 16)
+
+
+def normalize_datetime_fields(token_info):
+    for datetime_field in ['exp', 'iat', 'auth_time']:
+        if datetime_field not in token_info:
+            continue
+        token_info[datetime_field] = datetime.fromtimestamp(token_info[datetime_field], timezone.utc)
+    return token_info
+
+
+def parse_bearer_token(data, jwks=None, public_numbers=None):
+    validate_jwt_format(data)
+
+    data = to_bytes(data)
+    signing_input = data.rsplit(b".", 1)[0]
+    b64header_data, b64payload_data, b64signature = data.split(b".")
```

Comment:
Could this be
```suggestion
    signing_input, b64signature = data.rsplit(b".", 1)
    b64header_data, b64payload_data = signing_input.split(b".")
```
? Or am I missing something?

Reply:
It's equivalent, it doesn't achieves anything different?

Reply:
Yep, the main motivation was readability and semantical step separation, IIRC. Less chaining == easier to follow.